src/ray/gcs/gcs_server/gcs_health_check_manager.cc

```diff
@@ -129,15 +152,21 @@ void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck() {
   new (&context_) grpc::ClientContext();
   response_.Clear();
 
-  const auto deadline = now + absl::Milliseconds(manager_->timeout_ms_);
+  const auto deadline = now + absl::Milliseconds(manager->timeout_ms_);
   context_.set_deadline(absl::ToChronoTime(deadline));
+
+  // grpc and io_context contains two different eventloops, here we have to use shared
+  // pointer to make sure all memory accesses are valid.
   stub_->async()->Check(
-      &context_, &request_, &response_, [this, start = now](::grpc::Status status) {
+      &context_,
+      &request_,
+      &response_,
+      [this, start = now, manager = manager](::grpc::Status status) {
```

Comment:
we don't need to capture `manager` here. Since we captured `this`, we have the weak_ptr of the manager and we just need to do another `weak_ptr.lock()` inside the lambda.

Reply:
ok, https://github.com/ray-project/ray/pull/49493